Fix race condition in EventProcessor

wddgit opened the pull request:
#### PR description:

This fixes the problem from issue #26246. This is related
to a data race condition in EventProcessor. Almost always,
the correct data path wins the race and there are no
problems. In these cases, there should be no differences
in the output with or without this PR. But there are very rare
cases where the wrong code path wins the data race, then
endRun gets called late and this would most likely cause a
seg fault or very bad failure of some unpredictable nature. 

The condition is related to the deletion of the
LuminosityBlockProcessingStatus object. If we want
to call endRun after exiting processLumis, then
this status object must have already been destroyed
because it holds a shared_ptr to the RunResources
object that prevents endRun from running. If it isn't
destroyed then endRun may get called at some random
time much later which causes serious problems.

The writeT task in EventProcessor::globalEndLumiAsync
is the one that caused the failure in the issue. Although
there were several similar issues that this also fixes that could
result in the same problem.

#### PR validation:

It is hard to write a test for this, because it is a rare non reproducible problem. Even without the fix, almost always the right code path wins and there is absolutely no difference in how things execute in that case.

#### if this PR is a backport please specify the original PR:

Potentially I could backport this. These failures are very rare, so it is a judgement call whether it is worth the effort. Probably it was introduced when concurrent lumis were implemented.

wddgit committed:
Fix race condition in EventProcessor

First note that that vast majority of the time,
the correct code path wins the data race and there
is no problem, almost always. That is why we have
rarely seen failures related to this.

The condition is related to the deletion of the
LuminosityBlockProcessingStatus. If we want to
call endRun after exiting processLumis, then
this status object must have already been destroyed
because it holds a shared_ptr to the RunResources
object that prevents endRun from running. If it isn't
destroyed then endRun may get called at some random
time much later which causes all kinds of problems. Seg
faults would be likely.

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-26289/8976

- This PR adds an extra 36KB to repository

cmsbuild commented:
A new Pull Request was created by  @wddgit (W. David Dagenhart) for master.

It involves the following packages:

FWCore/Framework


@cmsbuild, @smuzaffar, @Dr15Jones can you please review it and eventually sign? Thanks.
@makortel, @Martin-Grunewald this is something you requested to watch as well.
@davidlange6, @slava77, @fabiocos you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

wddgit commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

Dr15Jones commented on FWCore/Framework/src/EventProcessor.cc:
```diff
@@ -1458,23 +1478,22 @@ namespace edm {
     FDEBUG(1) << "\tdeleteRunFromCache " << run << "\n";
   }
 
-  void EventProcessor::writeLumiAsync(WaitingTaskHolder task, std::shared_ptr<LuminosityBlockProcessingStatus> iStatus) {
```
What is stopping the LuminosityBlock from being deleted before the task created by writeLumiAsync finishes?

Dr15Jones replied:
Couldn't the shared_ptr be destroyed, the LuminosityBlock be returned the the ReusableObjectHolder and another lumi transition pick that object up and reset it while the writLumi is happening?

Dr15Jones replied:
Doesn't the LuminosityBlockProcessingStatus also prevent the LimitedTaskQueue from starting a new task? If the status object is deleted before writeLumi finishes, then another Lumi will start before we've finished with the one we are processing.

Dr15Jones replied:
I see now, the task passed into writeLumiAsync actually is keeping LuminosityBlockProcessingStatus alive. I withdraw my objection.

wddgit replied:
If you look in EventProcessor::globalEndLumiAsync at the first make_waiting_task (which is the one I did not modify), inside that task is a line "status.reset();". Until that point there is a shared_ptr pointer holding onto the status object somewhere and keeping it alive. Part of the time they are in a vector held as a data member by EventProcessor. A lot of other shared_ptr's are really unnecessary for keeping it alive. Some of what I did was remove some of the unnecessary shared_ptr so I did not have to worry about making sure they got deleted soon enough.

wddgit replied:
Note one thing that complicates this is that the destruction order of captured variables is not specified by the C++ standard. Same for function arguments.

Also note, sometimes I used a move on the shared_ptr so that it was already null when it hit the end of a function or lambda. I had to make a few modifications so I could do the moves.

Dr15Jones commented on FWCore/Framework/src/EventProcessor.cc:
```diff
@@ -1159,29 +1168,32 @@ namespace edm {
             beginGlobalTransitionAsync<Traits>(beginStreamsHolder,
                                                *schedule_,
-                                               *(status->lumiPrincipal()),
+                                               lumiPrincipal,
```
This line is just a clean up where an already existing local variable is used instead.

wddgit replied:
It is more than that. Because I moved status into beginStreamTask, it is already null at this point.

cmsbuild commented:
+1
Tested at: 6956ad75fc95e798001e89cfd94cbc2772c512cf
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-26289/33836/summary.html

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-26289/33836/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 8 differences found in the comparisons
 - DQMHistoTests: Total files compared: 32
 - DQMHistoTests: Total histograms compared: 3134059
 - DQMHistoTests: Total failures: 1
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3133861
 - DQMHistoTests: Total skipped: 197
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 31 files compared)
 - Checked 133 log files, 14 edm output root files, 32 DQM output files

wddgit commented:
Hold. I think I see a way to simplify this a little more so I will update part of the fix. Also, one more piece of information to consider if we are thinking about backporting this. LuminosityBlockProcessingStatus did not exist before release series 10_1_X, so this problem could not exist before that. And the problem is in 10_1_X. But again experience shows this is a very very rare problem ...

wddgit committed:
Simplify globalEndLumiAsync

Simply the function globalEndLumiAsync so it is more clear
to someone reading or maintaining the code who owns the lumi
status object. The status object is now moved into the first
task instead of copied. It is no longer moved into the second
task, which never really needed the shared_ptr in the first place.

I also edited the comments a bit and renamed a few things
to make this part of the code more clear.

This commit should not change behavior in any way.

cmsbuild commented:
The code-checks are being triggered in jenkins.

wddgit commented:
unhold

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-26289/8992

- This PR adds an extra 40KB to repository

wddgit commented:
please test

cmsbuild commented:
Pull request #26289 was updated. @cmsbuild, @smuzaffar, @Dr15Jones can you please check and sign again.

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: e644c6caaa21bbfb74db0e7cd620215a94ca315a
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-26289/33857/summary.html

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-26289/33857/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 4 differences found in the comparisons
 - DQMHistoTests: Total files compared: 32
 - DQMHistoTests: Total histograms compared: 3134059
 - DQMHistoTests: Total failures: 1
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3133861
 - DQMHistoTests: Total skipped: 197
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 31 files compared)
 - Checked 133 log files, 14 edm output root files, 32 DQM output files

Dr15Jones commented:
+1

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @davidlange6, @slava77, @smuzaffar, @fabiocos (and backports should be raised in the release meeting by the corresponding L2)

fabiocos commented:
+1
